jsbits/isomorphic.js

```diff
@@ -1,19 +1,40 @@
 window = typeof window === 'undefined' ? {} : window;
-window['copyDOMIntoVTree'] = function copyDOMIntoVTree(mountPoint, vtree, doc) {
+window['collapseSiblingTextNodes'] = function collapseSiblingTextNodes(vs) {
+  if (!vs) { return []; }
+  var ax = 0, adjusted = vs.length > 0 ? [vs[0]] : [];
+  for (var ix = 1; ix < vs.length; ix++) {
+    if (adjusted[ax]['type'] === 'vtext' && vs[ix]['type'] === 'vtext') {
+	adjusted[ax]['text'] += vs[ix]['text'];
+	continue;
+    }
+    adjusted[++ax] = vs[ix];
+  }
+  return adjusted;
+}
+
+window['copyDOMIntoVTree'] = function copyDOMIntoVTree(debug,mountPoint, vtree, doc) {
   if (!doc) { doc = window.document; }
   var node = mountPoint ? mountPoint.firstChild : doc.body.firstChild;
-  if (!window['walk'](vtree, node, doc)) {
-    console.warn('Could not copy DOM into virtual DOM, falling back to diff');
+    if (!window['walk'](debug,vtree, node, doc)) {
+    if (debug) {
+      console.warn('Could not copy DOM into virtual DOM, falling back to diff');
+    }
     // Remove all children before rebuilding DOM
-    while (node.firstChild)
-      node.removeChild(node.lastChild);
-    window['diff'](null, vtree, node, doc);
+    while (node.firstChild) node.removeChild(node.lastChild);
+    window['diff'](null, vtree, node.parentNode, doc);
     return false;
   }
+  if (debug) {
+    console.info ('Successfully prendered page');
+  }
   return true;
 }
 
-window['walk'] = function walk(vtree, node, doc) {
+window['diagnoseError'] = function diagnoseError(debug, vtree, node) {
+    if (debug) console.warn('VTree differed from node', vtree, node);
```

Comment:
I'm a bit skeptic about the `debug` variable here. The name of the variable (`debug`) does not immediately make clear *what* it debugs. Documentation has to be read.

Besides, what it does is show these debug prints. Is that something we actually expect end users to use? Is it easy enough to mismatch and hard enough to debug such mismatches? Biggest cause for mismatches AFAIK are those adjacent text nodes. That's exactly what you fix.

Perhaps we should look at a different means of debugging such mismatches, maybe some wrapper around the `miso` function?

Reply:
I've used `miso` for a couple of weeks and have run into multiple different causes for the mismatches, including:

* The problem you mention - adjacent text nodes
* Different Route URIs: #587
* Bug in the framework that renders `br_` differently: #585
* My own mistakes where I didn't ensure that the Model on the server is the same that the model I use in the client

Each time I had to resort to commenting out pieces of the code to narrow it down and before the fallback code was fixed in this pull request the app would just hang with no feedback.

I believe this functionality will remain useful, I have no strong opinions on the way to provide it.

Reply:
If we disagree, can we move the discussion to #588 and proceed with the main contents of this pull request? This is a really nice fix and I wouldn't want to stall it on a tangential issue.

Reply:
Alternatively, we could rename `debug` to `logLevel` and have multiple logging levels: `0 - off`, `1 - error`, `2 - warn`, `3 - info`, `4 - debug`, `5 - trace`.